src/DataStructures/DynamicBuffer.hpp

```diff
@@ -0,0 +1,138 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <vector>
+
+#include "DataStructures/DataVector.hpp"
+#include "Utilities/TypeTraits/GetFundamentalType.hpp"
+#include "Utilities/TypeTraits/IsIterable.hpp"
+
+namespace PUP {
+class er;
+}  // namespace PUP
+
+/**
+ * \ingroup DataStructuresGroup
+ *
+ * \brief A dynamically sized vector of `DataVector`s. For convenience it can
+ * also be instantiated for fundamental types.
+ *
+ * \details This class is useful when one wants to create a `std::vector<T>`
+ * with a size that is unknown at compile time. It allocates all `DataVector`s
+ * in a single memory chunk rather than allocating each individually. If the
+ * size of the vector is known at compile time, a `TempBuffer` object should be
+ * used instead.
+ *
+ * If needed it should be fairly straightforward to generalize to
+ * `ComplexDataVector`.
+ */
+
+template <typename T>
+class DynamicBuffer {
+ public:
+  static constexpr bool is_data_vector_type =
+      std::is_base_of_v<VectorImpl<tt::get_fundamental_type_t<T>, T>, T>;
+
+  DynamicBuffer() = default;
+
+  /*!
+   * Constructs a `DynamicBuffer`. The `number_of_vectors` corresponds to the
+   * number of `DataVector`s which are saved inside, each of which has size
+   * `number_of_grid_points`. `number_of_grid_points` has to be 1 if T is a
+   * fundamental type.
+   */
+  DynamicBuffer(const size_t number_of_vectors,
```

Comment:
The general advice is to avoid having inline constructors and large functions. Do you think the constructor and pup should be defined not inline? I'm not sure if this actually matters with class templates or if it's just classes

Reply:
I think with templated classes the only downside you might have is an increase in build time. Not sure if thats a valid argument but if you read the `if constexpr` branches separately the functions are only about 3 lines each. I dont know how stringent spectre is on this in general so I really cant say.

Reply:
It looks like in general we define these out-of-line, which is what I remember the advice being in Scott Meyer's book